src/Loop.php

```diff
@@ -17,20 +17,31 @@
     private static $driver = null;
 
     /**
-     * @var bool
+     * @var int
      */
-    private static $running = false;
+    private static $level = 0;
 
     /**
-     * Set the factory to be used to create a driver if none is passed to
-     * self::execute. A default driver will be created if none is running
-     * to support synchronous waits in traditional applications.
+     * Set the factory to be used to create a default drivers.
+     *
+     * Setting a factory is only allowed as long as no loop is currently running.
+     * Passing null will reset the default driver and remove the factory.
+     *
+     * The factory will be invoked if none is passed to Loop::execute. A default driver will be created to support
+     * synchronous waits in traditional applications.
      */
     public static function setFactory(LoopDriverFactory $factory = null)
     {
+        if (self::$level > 0) {
+            throw new \RuntimeException("Setting a new factory while running isn't allowed!");
+        }
+
         self::$factory = $factory;
 
-        if (!self::$running) {
+        if ($factory === null) {
+            self::$driver = null;
+            self::$registry = null;
+        } else {
             self::$driver = self::createDriver();
```

Comment:
Maybe `createDriver()` should be called lazily in `get()` instead of being called in `setFactory()`, in case it's expensive somehow? It was discussed that `setFactory()` could be called by loop implementation packages via composer.json's `autoload.files`, which could mean it sometimes gets called when the driver isn't actually going to be used.